plugins/Files/js/sagas/helpers.js

```diff
@@ -0,0 +1,161 @@
+import Path from 'path'
+import fs from 'fs'
+
+export const sendError = (e) => {
+	SiaAPI.showError({
+		title: 'Sia-UI Files Error',
+		content: e.toString(),
+	})
+}
+
+// siadCall: promisify Siad API calls.  Resolve the promise with `response` if the call was successful,
+// otherwise reject the promise with `err`.
+export const siadCall = (uri) => new Promise((resolve, reject) => {
+	SiaAPI.call(uri, (err, response) => {
+		if (err) {
+			reject(err)
+		} else {
+			resolve(response)
+		}
+	})
+})
+
+// Take a number of bytes and return a sane, human-readable size.
+export const readableFilesize = (bytes) => {
+	const units = ['PB', 'TB', 'GB', 'MB', 'KB', 'B']
+	let readableunit = 'B'
+	let readablesize = bytes
+	for (const unit in units) {
+		if (readablesize < 1000) {
+			readableunit = units[unit]
+			break
+		}
+		readablesize /= 1000
+	}
+	return readablesize.toFixed(2).toString() + ' ' + readableunit
+}
+
+// return a list of files filtered with path.
+// ... it's ls.
+export const ls = (files, path) => {
+	let fileList = files.filter((file) => file.uploadprogress >= 100)
+	fileList = fileList.filter((file) => file.siapath.indexOf(path) !== -1)
+	let parsedFiles = Map()
+	fileList.forEach((file) => {
+		let type = 'file'
+		const relativePath = file.siapath.substring(path.length, file.siapath.length)
+		let filename = Path.basename(relativePath)
+		let uploadprogress = Math.floor(file.uploadprogress)
+		if (relativePath.indexOf('/') !== -1) {
+			type = 'directory'
+			filename = relativePath.split('/')[0]
+			uploadprogress = ''
+		}
+		parsedFiles = parsedFiles.set(filename, {
+			size: readableFilesize(file.filesize),
+			name: filename,
+			siapath: file.siapath,
+			available: file.available,
+			uploadprogress: uploadprogress,
+			type,
+		})
+	})
+	return parsedFiles.toList().sortBy((file) => file.name)
+}
+
+// recursively version of readdir
+export const readdirRecursive = (path, files) => {
+	const dirfiles = fs.readdirSync(path)
+	let filelist
+	if (typeof files === 'undefined') {
+		filelist = List()
+	} else {
+		filelist = files
+	}
+	dirfiles.forEach((file) => {
+		const filepath = Path.join(path, file)
+		const stat = fs.statSync(filepath)
+		if (stat.isDirectory()) {
+			filelist = readdirRecursive(filepath, filelist)
+		} else if (stat.isFile()) {
+			filelist = filelist.push(filepath)
+		}
+	})
+	return filelist
+}
+
+// Parse a response from `/renter/downloads`
+// return a list of file downloads
+export const parseDownloads = (since, downloads) => {
+	let parsedDownloads = List(downloads).filter((download) => Date.parse(download.starttime) > since)
+	parsedDownloads = parsedDownloads.map((download) => ({
+		siapath: download.siapath,
+		name: Path.basename(download.siapath),
+		progress: Math.floor((download.received / download.filesize) * 100),
+		destination: download.destination,
+		type: 'download',
+		starttime: download.starttime,
+	}))
+	return parsedDownloads.sortBy((download) => -download.starttime)
```

Comment:
don't need to reassign the temp variable each time: `return List().filter().map().sortBy()`